### PR TITLE
fix(skillify): close test-pollution channel + self-heal stale lock-as-directory

### PR DESCRIPTION
**Pull request by efenocchi**
## Why this PR exists

`tests/claude-code/skillify-state.test.ts` used to write directly into the
developer's real `~/.deeplake/state/skillify/`. The "treats an unreadable
lock file as stale" case explicitly `mkdirSync`'d a directory at the lock
path inside the real home and cleaned it up with a swallow-error
`rmdirSync`. Any test run that crashed or was SIGKILL'd before the
finalizer left a `<key>.lock` directory behind. Over many test runs these
orphans accumulated (80+ on at least one dev machine) and poisoned every
production Stop-counter trigger that happened to hash to one of those
keys: `unlinkSync` on a directory throws `EISDIR`, the worker silently
returned `false`, no skill was ever mined for that project.

The bug is dormant on a machine that currently has zero orphans, but the
test fixture re-pollutes on every interrupted run. This PR closes the
pollution channel AND adds a self-heal path so machines that already have
orphans recover without manual `rm -rf`.

### Source

- **New `src/skillify/state-dir.ts`** — single source of truth for the
  state directory. Resolves `process.env.HIVEMIND_STATE_DIR` first,
  falls back to `~/.deeplake/state/skillify`. Computed lazily on every
  call so callers that swap the env between operations actually affect
  the path. Extracted into its own module to avoid an ESM cycle with
  `legacy-migration.ts`.

- **`src/skillify/state.ts`** — re-exports `getStateDir` from the new
  module. `tryAcquireWorkerLock` now:
  - accepts `EISDIR` (POSIX) and `EPERM` (Windows) on the failed
    `unlinkSync` so the self-heal fires cross-platform;
  - `lstatSync`-es the path as a cheap fast-path;
  - runs `rmdirSync(p)` for the actual recovery (shape-aware: a regular
    file at the path throws `ENOTDIR`, a non-empty dir throws
    `ENOTEMPTY`, neither destroys live state);
  - falls through to the atomic `openSync(p, "wx")` on every recovery
    error — that's the sole arbiter of who holds the lock, even when
    two processes both observe a stale dir at the same time.

- **`src/skillify/legacy-migration.ts`** — now resolves `current` via
  `getStateDir()` and derives `legacy` as the `skilify` sibling of
  `dirname(current)`. Under a `HIVEMIND_STATE_DIR=/tmp/...` redirect
  the sibling doesn't exist and the migration is a no-op (intended).
  The "already attempted" cache is keyed by resolved target dir so
  back-to-back env switches don't silently skip the migration.

- **`src/skillify/scope-config.ts`** — removed the module-level
  `STATE_DIR` const; `loadScopeConfig` + `saveScopeConfig` now resolve
  the path via `getStateDir()` per invocation. Without this, workers
  spawned after `HIVEMIND_STATE_DIR` was set would still write
  scope/team config into real home.

- **`src/skillify/manifest.ts`** — `manifestPath()` now returns
  `join(getStateDir(), "pulled.json")`. Same rationale.

- **`src/commands/skillify.ts`** — `hivemind skillify status` lists
  per-project state from `getStateDir()` instead of `homedir()`. Was
  the last split-brain reader.

### Tests

- `tests/claude-code/skillify-state.test.ts` redirects all state ops
  through `HIVEMIND_STATE_DIR=mkdtempSync(...)` in `beforeAll`,
  restores the prior env in `afterAll`, and uses `rmSync(..., {
  recursive: true, force: true })` for cleanup. Even a SIGKILL now
  only leaks debris into `/tmp`, which the OS reaps on boot.
- Replaced the legacy `expect(false)` edge-case test with three
  positive assertions:
  1. Self-heal recovers when the lock path is a stale empty dir.
  2. Self-heal does NOT destroy unexpected content inside a non-empty
     stale lock dir (`rmdirSync` bails on `ENOTEMPTY`).
  3. Self-heal does NOT rmSync a regular lock file that replaced the
     stale dir mid-recovery (TOCTOU regression).
- Added env-routing regressions:
  - `getStateDir()` resolves to the tmp dir; `migrateLegacyStateDir()`
    is a no-op in that world.
  - `saveScopeConfig` / `loadScopeConfig` / `manifestPath` all land
    inside `HIVEMIND_STATE_DIR`.

### Bundles

Regenerated agent bundles (claude-code, codex, cursor, hermes, pi) +
the CLI bundle to embed the env-aware getStateDir, the legacy
migration env routing, the rmdirSync self-heal, and the EPERM
acceptance.

## Adversarial review

Ran `/codex challenge` three rounds against this branch. Round 1
found a critical `legacy-migration.ts` bypass + scope-config/manifest
split + a TOCTOU double-acquire in the original recovery code. Round
2 found the `lstatSync` gate was not enough because `rmSync(...,
recursive:true, force:true)` still removes regular files. Round 3
found Windows uses `EPERM` instead of `EISDIR` and that `lstatSync`
errors were too conservative. Every finding above P1 was closed
before push; the remaining P2 (`openclaw/src/index.ts` has its own
inlined copy of the skillify state surface) is explicitly out of
scope — fixing that duplicated implementation is its own follow-up.

## Test plan

- [x] `npm run typecheck` clean
- [x] `npx vitest run` — 132 files, 2713 tests, all green (the 2
      pre-existing unhandled rejections in `session-notifications-hook`
      were verified to predate this branch via `git stash`)
- [x] `npm run build` — clean
- [x] End-to-end smoke: ran `claude -p "ok"` from this branch's
      project cwd, observed skillify worker spawn via SessionEnd,
      mine 22 conversation pairs across 3 sessions, run the gate,
      emit a SKIP verdict with concrete reasoning. Lock acquired
      and released cleanly.
- [ ] CI green
- [ ] CodeRabbit review

## Out of scope (called out for reviewers)

- `openclaw/src/index.ts` carries its own inlined copy of the
  skillify state / legacy migration / worker lock surface. It still
  hardcodes `homedir()` and has not received the self-heal. Fixing
  it requires touching the OpenClaw integration separately; tracked
  as a follow-up rather than expanding this PR.
- `package-lock.json` drift (`@emnapi/runtime`, `pg` as optionalDeps)
  and `codex/node_modules` are NOT included in this PR — those are
  unrelated to the lock-pollution fix.

## Release Notes

* **New Features**
  * Added support for configurable Skillify state directory via the `HIVEMIND_STATE_DIR` environment variable, allowing custom state storage locations.

* **Bug Fixes**
  * Improved robustness of legacy state migration to prevent redundant operations when using environment overrides.
  * Enhanced stale lock cleanup to better handle edge cases and prevent leftover artifacts from blocking future operations.

[Review Change Stack](https://app.coderabbit.ai/change-stack/activeloopai/hivemind/pull/181?utm_source=github_walkthrough&utm_medium=github&utm_campaign=change_stack)

### CODE REVIEW

**Comment by efenocchi** [comment]
@coderabbitai review

File: bundle/cli.js [hunk 1]
```diff
@@ -5086,27 +5086,33 @@ if (process.argv[1] && process.argv[1].endsWith("auth-login.js")) {
...
 // dist/src/skillify/legacy-migration.js
 import { existsSync as existsSync15, renameSync as renameSync3 } from "node:fs";
+import { dirname as dirname3, join as join19 } from "node:path";
+
+// dist/src/skillify/state-dir.js
 import { homedir as homedir8 } from "node:os";
 import { join as join18 } from "node:path";
+function getStateDir() {
+  return process.env.HIVEMIND_STATE_DIR ?? join18(homedir8(), ".deeplake", "state", "skillify");
+}
```
[the remark this answers is not in the document]
Comment (efenocchi) [reply bundle/cli.js:5239]: Fixed in b3376fc — `getStateDir()` now trims and checks length>0 before accepting the override, so `HIVEMIND_STATE_DIR=""` and `HIVEMIND_STATE_DIR="   "` both fall back to the home-based default. New regression test `empty/whitespace HIVEMIND_STATE_DIR is treated as unset` in `skillify-state.test.ts`. The bundle reflects the source fix as of the latest regenerate (5a57d48 source / 80c8c17 bundles).

File: bundle/cli.js [hunk 2]
```diff
@@ -5086,27 +5086,33 @@ if (process.argv[1] && process.argv[1].endsWith("auth-login.js")) {
...
+  const current = getStateDir();
+  if (attemptedFor.has(current))
     return;
-  attempted = true;
-  const root = join18(homedir8(), ".deeplake", "state");
-  const legacy = join18(root, "skilify");
-  const current = join18(root, "skillify");
+  attemptedFor.add(current);
+  const legacy = join19(dirname3(current), "skilify");
   if (!existsSync15(legacy))
```
[the remark this answers is not in the document]
Comment (efenocchi) [reply bundle/cli.js]: Fixed in 5a57d48 — `migrateLegacyStateDir()` now has a hard `if (process.env.HIVEMIND_STATE_DIR?.trim()) return;` guard at the top. The typo'd-sibling rename only runs against the canonical `~/.deeplake/state/` parent, never against an override-derived path. New regression test stages a coincidental `<state-parent>/skilify/` with a marker file and asserts it stays untouched when the env is set. Bundles regenerated in 80c8c17.

File: claude-code/bundle/capture.js [hunk 3]
```diff
@@ -998,25 +998,31 @@ function spawnSkillifyWorker(opts) {
...
 // dist/src/skillify/legacy-migration.js
 import { existsSync as existsSync5, renameSync as renameSync2 } from "node:fs";
+import { dirname as dirname4, join as join11 } from "node:path";
+
+// dist/src/skillify/state-dir.js
 import { homedir as homedir7 } from "node:os";
 import { join as join10 } from "node:path";
+function getStateDir() {
+  return process.env.HIVEMIND_STATE_DIR ?? join10(homedir7(), ".deeplake", "state", "skillify");
+}
```
[the remark this answers is not in the document]
Comment (efenocchi) [reply claude-code/bundle/capture.js:1167]: Fixed in b3376fc (same source change as the matching `bundle/cli.js` comment) — `getStateDir()` trims and checks length>0. Bundles regenerated.

File: claude-code/bundle/capture.js [hunk 4]
```diff
@@ -1183,8 +1188,22 @@ function tryAcquireWorkerLock(projectKey, maxAgeMs = 10 * 60 * 1e3) {
...
+        isDir = lstatSync(p).isDirectory();
+      } catch {
+      }
+      if (isDir) {
+        try {
+          rmdirSync(p);
+        } catch (rmErr) {
+          dlog3(`rmdir stale lock skipped for ${projectKey}: ${rmErr.message}`);
+        }
+      }
```
[the remark this answers is not in the document]
Comment (efenocchi) [reply claude-code/bundle/capture.js]: Fixed in b3376fc — `tryAcquireWorkerLock` now accepts `ENOENT` on the failed `unlinkSync` alongside `EISDIR` (POSIX) and `EPERM` (Windows). ENOENT means a racing process already cleared the stale path, so we fall through to the atomic `openSync(p, "wx")` which wins-or-loses as appropriate instead of returning false eagerly.

File: pi/bundle/skillify-worker.js [hunk 5]
```diff
@@ -561,25 +561,31 @@ function resolveRecordScope(args) {
...
+  const current = getStateDir();
+  if (attemptedFor.has(current))
     return;
-  attempted = true;
-  const root = join4(homedir4(), ".deeplake", "state");
-  const legacy = join4(root, "skilify");
-  const current = join4(root, "skillify");
+  attemptedFor.add(current);
+  const legacy = join5(dirname(current), "skilify");
   if (!existsSync3(legacy))
```
[the remark this answers is not in the document]
Comment (efenocchi) [reply pi/bundle/skillify-worker.js]: Fixed in 5a57d48 (same source change as the matching `legacy-migration.ts` comment) — explicit `HIVEMIND_STATE_DIR` set ⇒ `migrateLegacyStateDir()` returns immediately, no sibling stat, no rename. Bundles regenerated in 80c8c17.

File: src/skillify/legacy-migration.ts [hunk 6]
```diff
@@ -18,20 +38,19 @@
...
-  if (attempted) return;
-  attempted = true;
-  const root = join(homedir(), ".deeplake", "state");
-  const legacy = join(root, "skilify");
-  const current = join(root, "skillify");
+  const current = getStateDir();
+  if (attemptedFor.has(current)) return;
+  attemptedFor.add(current);
+  const legacy = join(dirname(current), "skilify");
   if (!existsSync(legacy)) return;
```
[the remark this answers is not in the document]
Comment (efenocchi) [reply src/skillify/legacy-migration.ts:70]: Fixed in 5a57d48 — added an explicit `if (process.env.HIVEMIND_STATE_DIR?.trim()) return;` guard at the top of `migrateLegacyStateDir()`. The legacy-typo migration is meaningful only at the canonical home-based default; any caller that overrides the state dir has its own path lifecycle and should never trigger a sibling rename. Regression test `migrateLegacyStateDir is a hard no-op when HIVEMIND_STATE_DIR is set, even if a sibling 'skilify' dir coincidentally exists` covers it.

File: src/skillify/state-dir.ts [hunk 7]
```diff
@@ -0,0 +1,29 @@
...
+ * subsystem at a `mkdtempSync()` directory without monkey-patching
+ * `os.homedir`. Falls back to `~/.deeplake/state/skillify`.
+ */
+
+import { homedir } from "node:os";
+import { join } from "node:path";
+
+export function getStateDir(): string {
+  return process.env.HIVEMIND_STATE_DIR ?? join(homedir(), ".deeplake", "state", "skillify");
+}
```
[the remark this answers is not in the document]
Comment (efenocchi) [reply src/skillify/state-dir.ts:39]: Fixed in b3376fc — `getStateDir()` now reads `process.env.HIVEMIND_STATE_DIR?.trim()` and only returns it when length > 0, otherwise falls back to the homedir-based default. Test `empty/whitespace HIVEMIND_STATE_DIR is treated as unset` exercises both `""` and `"   "`.

File: tests/claude-code/skillify-state.test.ts [hunk 8]
```diff
@@ -231,21 +269,127 @@ describe("worker lock", () => {
...
+    expect(getStateDir()).toBe(STATE_DIR);
+    expect(getStateDir().startsWith(homedir())).toBe(false);
+
+    const fs = require("node:fs");
+    const tmpLegacy = join(STATE_DIR, "..", "skilify");
+    expect(fs.existsSync(tmpLegacy)).toBe(false);
+    expect(() => migrateLegacyStateDir()).not.toThrow();
+    // Did not magic the legacy sibling into existence either.
+    expect(fs.existsSync(tmpLegacy)).toBe(false);
+  });
```
[the remark this answers is not in the document]
Comment (efenocchi) [reply tests/claude-code/skillify-state.test.ts:381]: Fixed in b3376fc — the legacy-migration env-routing test now calls `vi.resetModules()` before dynamically importing both `state-dir.js` and `legacy-migration.js`, so the module-level `attemptedFor` Set is genuinely empty and the function body actually runs. The `empty/whitespace` test added in the same commit follows the same pattern.

**Comment by efenocchi** [comment]
## CodeRabbit review status

All 8 inline findings + the 1 outside-diff finding are addressed. Replies posted on each inline comment with the specific commit.

| # | File | Concern | Commit |
|---|------|---------|--------|
| 1 | `bundle/cli.js:5106` | Empty/whitespace `HIVEMIND_STATE_DIR` accepted as live root | `b3376fc` |
| 2 | `bundle/cli.js:5117` | Sibling rename on override-derived path | `5a57d48` |
| 3 | `claude-code/bundle/capture.js:1016` | Empty env duplicate of #1 | `b3376fc` |
| 4 | `claude-code/bundle/capture.js:1207` | ENOENT on unlinkSync returned false instead of falling through | `b3376fc` |
| 5 | `pi/bundle/skillify-worker.js:590` | Sibling rename duplicate of #2 | `5a57d48` |
| 6 | `src/skillify/legacy-migration.ts:56` | Source of #2/#5 | `5a57d48` |
| 7 | `src/skillify/state-dir.ts:39` | Source of #1/#3 | `b3376fc` |
| 8 | `tests/.../skillify-state.test.ts:381` | Test masked by module-level cache | `b3376fc` |
| 9 (outside-diff) | `cursor/bundle/skillify-worker.js:583` | `renameSync` race codes (ENOENT/EEXIST/ENOTEMPTY) rethrown | `b3376fc` |

Bundles regenerated in `a20bde0` (for #1/#3/#4/#7/#8/#9 fixes) and `80c8c17` (for #2/#5/#6 fix).

New regression tests in `tests/claude-code/skillify-state.test.ts`:
- `empty/whitespace HIVEMIND_STATE_DIR is treated as unset` (covers #1/#3/#7)
- `migrateLegacyStateDir is a hard no-op when HIVEMIND_STATE_DIR is set, even if a sibling 'skilify' dir coincidentally exists` (covers #2/#5/#6)
- `vi.resetModules()` before the `legacy-migration` import in the env-routing test (covers #8)

`npm run audit:openclaw -- --criticals-only` reports 0 critical (the 2 pre-existing potential-exfiltration warns are non-blocking with the `--criticals-only` flag the CI release gate uses, and predate this branch). Full vitest run: 132 files, 2715 tests green.

File: src/skillify/legacy-migration.ts [hunk 9]
```diff
@@ -8,39 +8,79 @@
...
+  // in the canonical home-based layout. The `skilify`-vs-`skillify`
+  // typo migration is a one-shot upgrade against the historical
+  // `~/.deeplake/state/` parent and has no meaning anywhere else.
+  // Without this guard, an override like `HIVEMIND_STATE_DIR=/tmp/foo`
+  // would still cause us to `existsSync('/tmp/skilify')` and — if some
+  // unrelated tool happened to have created that directory —
+  // `renameSync('/tmp/skilify', '/tmp/foo')` and move someone else's
+  // content into our state path. Tests (and any other caller that
+  // overrides) get a hard no-op here and a clean tmp-dir start.
+  if (process.env.HIVEMIND_STATE_DIR?.trim()) return;
```
Comment (kaghni) [inline src/skillify/legacy-migration.ts:64]: if it's not used anywhere you can just delete it with usage

Comment (kaghni) [reply src/skillify/legacy-migration.ts:64]: @coderabbitai is that the case?

Comment (efenocchi) [reply src/skillify/legacy-migration.ts:64]: `attemptedFor` is used inside `migrateLegacyStateDir()` itself at lines 65-66 — the read+write Set is what makes the migration idempotent across the 8 callers that invoke `migrateLegacyStateDir()` at the start of every public state operation:

```
src/skillify/state.ts:140        readState
src/skillify/state.ts:151        writeState
src/skillify/state.ts:160        withRmwLock
src/skillify/state.ts:268        tryAcquireWorkerLock
src/skillify/scope-config.ts:47  loadScopeConfig
src/skillify/scope-config.ts:71  saveScopeConfig
src/skillify/manifest.ts:77      loadManifest
src/skillify/manifest.ts:139     saveManifest
```

So it can't be deleted outright. That said, CodeRabbit's point is right: with the `HIVEMIND_STATE_DIR` guard at the top of the function, `getStateDir()` always resolves to the same path within a process, so the Set will only ever hold one entry. Downgrading to a plain `let attempted = false` is a clean simplification — pushing that fix now.

Comment (efenocchi) [reply src/skillify/legacy-migration.ts:64]: Downgraded to a plain `let attempted = false` in 9bfb4f5. Same behavior, 3 lines lighter. Full test suite (2727 tests) still green.

Comment (kaghni) [reply src/skillify/legacy-migration.ts:64]: @coderabbitai summarize the thread briefly

**Review by kaghni — APPROVED** [review]
